packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs

```diff
@@ -4,7 +4,45 @@
 mod entity;
 mod link;
 
+use serde::{Deserialize, Serialize};
+use utoipa::Component;
+
 pub use self::{
     entity::{Entity, EntityId, PersistedEntity, PersistedEntityIdentifier},
     link::Link,
 };
+use crate::{
+    ontology::{
+        PersistedDataType, PersistedEntityType, PersistedLinkType, PersistedPropertyType,
+        QueryDepth,
+    },
+    store::query::Expression,
+};
+
+/// Query to read [`Entities`], which are matching the [`Expression`].
+///
+/// [`Entities`]: Entity
+#[derive(Debug, Deserialize, Component)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct EntityQuery {
+    #[serde(rename = "query")]
+    pub expression: Expression,
+    #[component(value_type = number)]
+    pub data_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub property_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub link_type_query_depth: QueryDepth,
+    #[component(value_type = number)]
+    pub entity_type_query_depth: QueryDepth,
```

Comment:
This is making me wonder if we actually want to include the depth parameters within the expression.. Especially if we have nested queries, I think you probably want to be able to specify the type parameters alongside the type expression, and _not_ have to provide type parameters if you're not supplying a type portion of the expression.

But I can also see value in only having to define them once, globally, for the query. 🤔

Reply:
I think I cannot follow your arguments. What do you mean by nested queries? We requesting entities only here. The expression only filters, _which_ entities we are searching for. E.g. having a query like "Give me all entities, which has a link of type X" would still return the list of entities, not the links (unless specified by the query depth).

Reply:
Have spoken through this offline, persisting it here would be too verbose (apologies if anyone ever finds this comment)